### PR TITLE
blocks: align output tags in selector block

### LINKED ISSUE #3569: runtime/blocks (?): tag propagation anomalies when selectively dropping data
While debugging a simple block, we noticed some unusual tag propagation behavior when dropping large amounts of data. This can be demonstrated using the in-tree selector block as follows:

image

Initially, the tags are correct:

image

However, by selecting and deselecting the `Enabled` check box a couple times, its clear the tag propagation has some errors:

image

This example was put together with only in-tree blocks, but using the `tag_uhd_offset_f` block from gr-timing_utils, its possible to see that old tags remain in the output buffer with newer data (runtime does not account for the updated tag placement in the output buffer now that some data has been dropped?).

We attributed this to a runtime issue, as it appears that runtime appears to have sufficient information to propagate tags correctly (consuming X samples, producing 0 samples), though this is a bit of a special case of `general_work()` and propagation can also be fixed in the block directly via manual tag propagation in the selector block.

Regardless, tags should propagate correctly through this block. Thoughts on the correct approach?

This affects 3.7 (after the new selector was added), 3.8, and master. Flowgraph is here: [tagpropagation.zip](https://github.com/gnuradio/gnuradio/files/4774544/tagpropagation.zip)

### PR DESCRIPTION
**Pull request by mormj**
Aligns the output tags using nitems_written as recommended by @rear1019 in #3572 to address the situation where tags on the input are not aligned with the output due to dropped samples

@mrjacobagilbert @rear1019 - please take a look, I believe this can supercede #3572 with less bookkeeping

Fixes #3569

### CODE REVIEW

File: gr-blocks/lib/selector_impl.cc [hunk 1]
```diff
@@ -149,8 +151,20 @@ int selector_impl::general_work(int noutput_items,
...
     uint8_t** out = (uint8_t**)&output_items[0];
 
+
     gr::thread::scoped_lock l(d_mutex);
     if (d_enabled) {
+        auto nread = nitems_read(d_input_index);
+        auto nwritten = nitems_written(d_output_index);
+
+        std::vector<tag_t> tags;
+        get_tags_in_window(tags, 0, 0, noutput_items);
```
Comment (rear1019) [inline gr-blocks/lib/selector_impl.cc:161]: This selects tags from the first input. Why not from input `d_input_index`?

File: gr-blocks/lib/selector_impl.cc [hunk 2]
```diff
@@ -149,8 +151,20 @@ int selector_impl::general_work(int noutput_items,
...
+
+        for (auto tag : tags) {
+            tag.offset -= (nread - nwritten);
```
Comment (rear1019) [inline gr-blocks/lib/selector_impl.cc:164]: This should be correct, however, I can not test right now.

Me suggestion from #3572 to simply use `tag.offset = nitems_written(d_output_index)` is wrong. The tag offset must be modified to `offset at input − number of dropped samples = offset at input − (number items read − number items written)` (as used in this code). Or put another way to `number items written + relative offset at input = number items written + (offset at input − number items read)`.

**Review by rear1019 — APPROVED** [review]

**Comment by mrjacobagilbert** [comment]
@mormj I took a look over this and agree this is a better solution than #3569. i tested it using the attached FG and it looks like it behaves properly.

[tagpropagation.zip](https://github.com/gnuradio/gnuradio/files/5137817/tagpropagation.zip)

**Comment by mormj** [comment]
Thanks @mrjacobagilbert and @rear1019 - merging based on review

**Pull request merged**